### PR TITLE
catch any remaining exception then call APP_ABORT

### PR DESCRIPTION
**Pull request by PDoakORNL**
Ideally exceptions would be handled closer to throw location where useful information could be added.

### CODE REVIEW

File: src/QMCApp/qmcapp.cpp [hunk 1]
```diff
@@ -43,205 +44,220 @@ void output_hardware_info(Communicate* comm, Libxml2Document& doc, xmlNodePtr ro
...
-  myinput             = myinput.substr(0, myinput.size() - 4);
-  logname << myinput;
-
-  if (qmcComm->rank() != 0)
-  {
-    outputManager.shutOff();
-    // might need to redirect debug stream to a file per rank if debugging is enabled
   }
-  if (inputs.size() > 1 && qmcComm->rank() == 0)
+  catch (const std::exception& e)
```
Comment (markdewing) [inline src/QMCApp/qmcapp.cpp]: Does std::bad_alloc give a reasonable message in what(), or should there be a separate catch for the out-of-memory condition?

Comment (PDoakORNL) [reply src/QMCApp/qmcapp.cpp]: Well what() is all you have to call.  What is in it is implementation defined I believe.

**Comment by ye-luo** [comment]
Can someone verify in #1991 the code actually throws an std::bad_alloc and it can be properly catched. If it does, I do believe we can move the try catch to the clone function and provide more helpful message. We only allocate vast memory in very few places and can provide very specific message.

**Comment by PDoakORNL** [comment]
So I'm going to try and do this in a VM, Isn't a very nice thing to try and do to HPC a node.

**Comment by prckent** [comment]
We already have exceptions so adding this catch seems important. Any objections?

**Comment by ye-luo** [comment]
More precise, local and verbose catches are preferred than a global one.

**Comment by prckent** [comment]
But is there a problem to add this in the interim? Are we closing this PR?
@PDoakORNL what is your preference?

**Comment by jtkrogel** [comment]
Potentially dangerous if any raised exceptions are actually innocuous or incorrectly raised.  Broad scope includes all libraries too, right?  This is speculative, but will be interesting to see if the code becomes fragile due to this.

**Comment by PDoakORNL** [comment]
An unhandled exception will currently cause the code to terminate without calling app_abort

**Comment by prckent** [comment]
@jtkrogel The code is already fragile due to this. This PR will fix the problem.

**Comment by ye-luo** [comment]
I'd like to see a confirmation that the catch works as we expected. @PDoakORNL you can simply put a large meshfactor in the input file to claim huge memory.

**Comment by ye-luo** [comment]
@PDoakORNL is the major concern that std::terminate instead off MPI_ABORT may cause program  hangling?

**Comment by PDoakORNL** [comment]
Yes

**Comment by PDoakORNL** [comment]
ok to test

**Comment by prckent** [comment]
Merging since the ALCF CI appears to have technical problems and not issues with this PR.

**Pull request merged**